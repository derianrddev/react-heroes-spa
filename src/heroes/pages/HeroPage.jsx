import { useMemo } from "react";
import { Navigate, useNavigate, useParams } from "react-router-dom"
import { getHeroById } from "../helpers";

export const HeroPage = () => {

  const { heroId } = useParams();
  const navigate = useNavigate();

  const hero = useMemo( () => getHeroById( heroId ), [ heroId ]);

  const onNavigateBack = () => {
    navigate(-1);
  }

  if (!hero) {
    return <Navigate to="/marvel" />;
  }

  const { id, superhero, publisher, alter_ego, first_appearance, characters } = hero;
  const heroImageUrl = `/assets/heroes/${ id }.jpg`;

  return (
    <>
      <div className="row my-5">
        <div className="col-4">
          <img
           className="img-thumbnail animate__animated animate__fadeInLeft" 
           src={ heroImageUrl } 
           alt={ superhero } 
          />
        </div>
        <div className="col-8">
          <h3>{ superhero }</h3>
          <ul className="list-group list-group-flush">
            <li className="list-group-item"><b>Alter ego:</b> { alter_ego }</li>
            <li className="list-group-item"><b>Publisher:</b> { publisher }</li>
            <li className="list-group-item"><b>First appearance:</b> { first_appearance }</li>
          </ul>
          <h5 className="mt-3"> Characters </h5>
          <p>{ characters }</p>
          <button
           className="btn btn-outline-primary" 
           onClick={ onNavigateBack }
          >
            Exit
          </button>
        </div>
      </div>
    </>
  )
}
