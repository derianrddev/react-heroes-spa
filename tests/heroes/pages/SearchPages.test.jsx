import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, useNavigate } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate
}));

describe('Pruebas en <SearchPages />', () => {

  beforeEach( () => jest.clearAllMocks() );

  test('Debe de mostrarse correctamente con valores por defecto', () => { 

    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    )
    expect( container ).toMatchSnapshot();

  });

  test('Debe de mostrar a Batman y el input con el valor del queryString', () => { 

    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchPage />
      </MemoryRouter>
    )
    const input = screen.getByRole('textbox');
    expect( input.value ).toBe('batman');
    const image = screen.getByRole('img');
    expect( image.src ).toContain('dc-batman.jpg');
    const alert = screen.getByLabelText('alert-danger');
    expect( alert.style.display ).toBe('none');

  });

  test('Debe de mostrar un error si no se encuentra el hÃ©roe (batman123)', () => { 

    render(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <SearchPage />
      </MemoryRouter>
    )
    const alert = screen.getByLabelText('alert-danger');
    expect( alert.style.display ).toBe('block');

  });

  test('Debe de llamar el navigate a la pantalla nueva', () => { 

    render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    )
    const input = screen.getByRole('textbox'); 
    const form = screen.getByRole('form'); 
    fireEvent.change( input, { name: 'searchText', target: { value: 'batman'} } );
    fireEvent.submit( form );
    expect( mockedUseNavigate ).toHaveBeenCalledWith('?q=batman');

  });

});