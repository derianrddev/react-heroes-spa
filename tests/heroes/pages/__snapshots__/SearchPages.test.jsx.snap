// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPages /> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <div
    class="row mt-5"
  >
    <div
      class="col-5"
    >
      <h4>
        Searcher
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
        style="display: block;"
      >
        Search a hero
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        There's not a hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
